package/ios/RNSkia-iOS/SkiaMetalSurfaceFactory.h

```diff
@@ -0,0 +1,27 @@
+#import <MetalKit/MetalKit.h>
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdocumentation"
+
+#import <include/gpu/GrDirectContext.h>
+
+#pragma clang diagnostic pop
+
+using SkiaMetalContext = struct SkiaMetalContext {
+  id<MTLCommandQueue> commandQueue = nullptr;
+  sk_sp<GrDirectContext> skContext = nullptr;
+};
+
+class ThreadContextHolder {
+public:
+  static thread_local SkiaMetalContext ThreadMetalContext;
+};
+
+class SkiaMetalSurfaceFactory {
+private:
+  static id<MTLDevice> device;
+
+public:
+  static SkiaMetalContext &createSkiaDirectContextIfNecessary();
```

Comment:
makeWindowedSurface doesn't work because we need to present. We could do:
```
class SkiaMetalSurfaceFactory {
private:
  static id<MTLDevice> device;
  static SkiaMetalContext &createSkiaDirectContextIfNecessary();

public:
  static bool drawOnScreen(CAMetalLayer *layer, const std::function<void(SkCanvas *)> &cb);
  static sk_sp<SkSurface> makeOffscreenSurface(int width, int height);
};
```

with this API, making createSkiaDirectContextIfNecessary() private works.

Reply:
done 👍

Reply:
I think we should do makeWindowedSurface in the factory as suggested, present is already done using the drawable which is kept in the canvas provider - this is a suggested solution in the canvas provider:

```cpp
    id<CAMetalDrawable> currentDrawable = [_layer nextDrawable];
    if (currentDrawable == nullptr) {
      return false;
    }

    // This is new:
    auto surface = SkiaMetalSurfaceFactory::makeWindowedSurface(_layer.drawableSize.width,
       _layer.drawableSize.height, currentDrawable.texture);

    SkCanvas *canvas = skSurface->getCanvas();
    cb(canvas);

    skSurface->flushAndSubmit();

    id<MTLCommandBuffer> commandBuffer(
        [metalContext.commandQueue commandBuffer]);
    [commandBuffer presentDrawable:currentDrawable];
    [commandBuffer commit];
```

This way we get better symmetry in the factory - and we move some of the more special Skia/Metal code into the surface factory.

I think it makes sense?

Reply:
I like it but in the code snippet above the skiaMetalContext needs to be accessed for the commandBuffer

Reply:
Right - good point - what about returning something like what we did in Android implementation, like the `WindowSurfaceHolder`?

Reply:
But the `drawOnScreen` shouldn't be part of the surface class?